compiler/qsc_frontend/src/resolve.rs

```diff
@@ -707,18 +738,30 @@ fn resolve_scope_locals(
 
     None
 }
+#[derive(PartialEq, Hash, Eq, Debug)]
+/// Represents the resolution of implicit opens, but also
+/// retains the namespace that the resolution comes from.
+/// This retained namespace string is used for error reporting.
+struct ImplicitOpenResolution<'a> {
+    /// the namespace that the `result` comes from.
+    namespace: &'a str,
+    /// the resolution that results from looking for a specific name in `namespace`.
+    result: Res,
+}
 
-fn resolve_implicit_opens(
+fn resolve_implicit_opens<'a, 'b>(
     kind: NameKind,
-    globals: &GlobalScope,
-    namespaces: impl IntoIterator<Item = impl AsRef<str>>,
-    name: &str,
-) -> HashSet<Res> {
+    globals: &'b GlobalScope,
+    namespaces: impl IntoIterator<Item = &'a &'a str>,
+    name: &'b str,
+) -> HashSet<ImplicitOpenResolution<'a>> {
```

Comment:
The uniqueness is based only on `Res`, and the namespace name is just metadata. This type allows you to return the same `Res` multiple times if the namespace name is different, which is impossible.

That's why `resolve_explicit_opens` returns a `HashMap<Res, &'a Open>` - here it could be `HashMap<Res, &'a str>` (or if you prefer to newtype the string, the struct should contain just the string, not the `Res`).